File: src/wordle/WordleRulesController.java
package wordle;

final public class WordleRulesController {
    private final Dictionary dictionary = new Dictionary();
    private final Word solution;
    private final CandidateSolutions candidateSolutions;

    public WordleRulesController(
            final int totalPossibleLetters,
            final int totalPossibleGuesses) {
        solution = dictionary.chooseSolution(totalPossibleLetters);
        candidateSolutions = new CandidateSolutions(totalPossibleGuesses);
    }

    public boolean isGameOver() {
        return hasGuessedTheCorrectWord() || !candidateSolutions.hasAnotherTurn();
    }

    private boolean hasGuessedTheCorrectWord() {
        return candidateSolutions.has(solution);
    }

    public CandidateSolutions getCandidateSolutions() {
        return candidateSolutions;
    }

    public void guess(final Word candidateWord) {
        candidateSolutions.guess(candidateWord);
    }
}

File: src/wordle/WordleRulesController.kt
package wordle

class WordleRulesController(
        val totalPossibleLetters: Int,
        totalPossibleGuesses: Int) {
    private val dictionary = Dictionary()
    private val solution: Word = dictionary.chooseSolution(totalPossibleLetters)
    val candidateSolutions: CandidateSolutions = CandidateSolutions(totalPossibleLetters, totalPossibleGuesses)

    val isGameOver: Boolean
        get() = hasGuessedTheCorrectWord() || !candidateSolutions.hasAnotherTurn()

    private fun hasGuessedTheCorrectWord(): Boolean {
        return candidateSolutions.has(solution)
    }

    fun guess(theGuess: CharArray) {
        candidateSolutions.guess(solution, theGuess)
    }
}